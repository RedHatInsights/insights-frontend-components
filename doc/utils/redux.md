# Redux
We provide some improvements over redux so projects using this package can use redux to it's full.

### Middleware listener
Because there are places where creating new reducer is too robust and might lead to wrong state mutation there is middleware which can create listeners on redux actions. Anytime new action is triggered it looks trough this set of listeners and calls callbacks.

#### Usage of middleware listener
```javascript
import { applyMiddleware } from 'redux';
import { MiddlewareListener } from '@red-hat-insights/insights-frontend-components';
const listenerMiddleware = new MiddlewareListener();
this.store = createStore(
    reducers,
    initState,
    applyMiddleware(listenerMiddleware.getMiddleware())
);
```

After creating such listener middleware you can simply call `addNew` method with type over which you want to listen and callback that is called when such action is called on redux.

addNew params:
* on - type of action to listen to
* callback - function that is called when action is triggered

Callback params:
* data - payload from action
* preventBubble - function to prevent bubbling of such redux event. New event is triggered with type `@@config/action-stopped` and payload as action that was stopped.
```javascript
const removeListener = listenerMiddleware.addNew({on: 'someType', (event) => event.preventBubble()})
```

Please clean after yourself by calling function which is returned from `listenerMiddleware`.
### Reducer registery
To enable registering reducers on the fly you may use reducer registery. There are also helper functions to help you improve your work with redux.

#### Usage of reducer registery
Simply create new instance of reducer registery, you may wish to supply some parameters while creating new store so you 
can change behaviour of your state.

Parameters:
* `initState` - default to empty object
* `middlewares` - default to empty array
* `composeEnhancers` - default to `compose` from redux library

```javascript
import { ReducerRegistry } from '@red-hat-insights/insights-frontend-components';
import { middlewares } from './middlewareSettings';

const reduxRegistry = new ReducerRegistry(initState, middlewares);
```

Available methods:
* `getStore` - to dispatch or to read state of store
* `register` - to add multiple reducers

### Helper function
#### `applyReducerHash`
If you don't like switch statements in your reducers you may use object literals, this function will map such object to
reducer
```javascript
const SOME_ACTION = 'some-action';
function reducerFunction(state, action) {
    return {
        ...state,
        someProps: action.payload
    };
}

const customReducer = applyReducerHash({
    [SOME_ACTION]: reducerFunction
})
```

Later you can assign this `customReducer` to change the store by calling `register` function.

#### `dispatchActionsToStore`
If application has some actions which needs to be dispatched over multiple stores or they can't be connected via `redux-redux` it's convenient to use this function.

It's as easy as calling dispatchActionsToStore with multiple actions (either trough object or array) and store over which you want to dispatch these functions.
```javascript
import * as actions from './actions';
import store from './singleton-store';
import { dispatchActionsToStore } from '@red-hat-insights/insights-frontend-components';

export default dispatchActionsToStore(actions, store);
```
