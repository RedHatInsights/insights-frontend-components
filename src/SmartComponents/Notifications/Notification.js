import React, { Component } from 'react';
import { Alert, Button } from '@patternfly/react-core';
import { CloseIcon } from '@patternfly/react-icons';
import PropTypes from 'prop-types';
import './notifications.scss';

/**
 * Add some enter and dismiss animation later when PF has designs
 */

const DEFAULT_DELAY = 5000;

class Notification extends Component {
    constructor(props) {
        super(props);
        this.handleDismiss = this.handleDismiss.bind(this);
        if (!props.dismissable) {
            this.dismissTimeout = setTimeout(() => this.handleDismiss(), props.dismissDelay);
        }
    }

    handleDismiss() {
        this.props.onDismiss(this.props.id);
    }

    componentWillUnmount() {
        if (this.dismissTimeout) {
            clearTimeout(this.dismissTimeout);
        }
    }

    render() {
        const { description, dismissable, onDismiss, dismissDelay, ...rest } = this.props;
        return (
            <Alert
                className="notification-item"
                { ...rest }
                action={ dismissable ?
                    <Button
                        aria-label="close-notification"
                        variant="plain"
                        onClick={ this.handleDismiss }
                    >
                        <CloseIcon/>
                    </Button> : null
                }
            >
                { description }
            </Alert>
        );
    }
}

Notification.propTypes = {
    dismissable: PropTypes.bool,
    onDismiss: PropTypes.func.isRequired,
    id: PropTypes.string.isRequired,
    variant: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    dismissDelay: PropTypes.number
};

Notification.defaultProps = {
    dismissDelay: DEFAULT_DELAY
};

export default Notification;
