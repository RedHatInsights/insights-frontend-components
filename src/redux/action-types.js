const asyncInventory = [
    'LOAD_ENTITIES',
    'LOAD_ENTITY'
].reduce((acc, curr) => [
    ... acc,
    ...[ curr, `${curr}_PENDING`, `${curr}_FULFILLED`, `${curr}_REJECTED` ]
], []);

export const INVENTORY_ACTION_TYPES = [
    ...asyncInventory
]
.reduce((acc, curr) => {
    acc[curr] = curr;
    return acc;
},
{}
);

export const ACTION_TYPES = [
    ...asyncInventory
]
.reduce((acc, curr) => {
    acc[curr] = curr;
    return acc;
},
{}
);

export const SELECT_ENTITY = 'SELECT_ENTITY';
export const CHANGE_SORT = 'CHANGE_SORT';
export const FILTER_ENTITIES = 'FILTER_ENTITIES';
export const APPLICATION_SELECTED = 'APPLICATION_SELECTED';
