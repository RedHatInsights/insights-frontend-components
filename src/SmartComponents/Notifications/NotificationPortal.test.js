import React from 'react';
import { mount, shallow, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store'
import { NotificationsPortal } from './';
import Notification from './Notification';
import { REMOVE_NOTIFICATION } from '../../redux/action-types';

describe('Notification portal', () => {
  let initialProps;
  let middlewares;;
  let mockStore;
  beforeEach(() => {
    initialProps = {}
    middlewares = [];
    mockStore = configureStore(middlewares);
    initialProps = {
      store: mockStore({})
    }
  });

  it('should return no component when no notifications given', () => {
    const t = () => {
      try {
        new ReactWrapper(mount(<NotificationsPortal {...initialProps} />))
      } catch (error) {
        throw new TypeError
      }
    }
    expect(t).toThrow(TypeError);
  });

  it('should render notifications given from store', () => {
    const modifiedStore = mockStore({
      notifications: [{
        id: '0',
        variant: 'success',
        title: 'Notification title',
        description: 'Some meaningfull description',
        dismissable: true
      }]
    })
    const wrapper = shallow(<NotificationsPortal {...initialProps} store={modifiedStore}/>).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render notifications given as direct props', () => {
    const wrapper = shallow(<NotificationsPortal {...initialProps} notifications={[{
      id: '0',
      variant: 'success',
      title: 'Notification title',
      description: 'Some meaningfull description',
      dismissable: true
    }]} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render notifications given as direct props over store notifications', () => {
    const modifiedStore = mockStore({
      notifications: [{
        id: 'store notifications',
        variant: 'success',
        title: 'Notification title',
        description: 'Some meaningfull description',
        dismissable: true
      }]
    })
  
    const wrapper = shallow(<NotificationsPortal {...initialProps} store={modifiedStore} notifications={[{
      id: 'Direct notification',
      variant: 'success',
      title: 'Notification title',
      description: 'Some meaningfull description',
      dismissable: true
    }]} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call default removeNotification function when none given', () => {
    const modifiedStore = mockStore({
      notifications: [{
        id: 'store notifications',
        variant: 'success',
        title: 'Notification title',
        description: 'Some meaningfull description',
        dismissable: true
      }]
    })
    const wrapper = mount(<NotificationsPortal {...initialProps} store={modifiedStore}/>)
    wrapper.find(Notification).find('button').simulate('click');
    expect(modifiedStore.getActions()).toEqual([
      expect.objectContaining({
        type: REMOVE_NOTIFICATION
      })
    ]);
  });

  it('should call custom removeNotification function', () => {
    const modifiedStore = mockStore({
      notifications: [{
        id: 'store notification',
        variant: 'success',
        title: 'Notification title',
        description: 'Some meaningfull description',
        dismissable: true
      }]
    })
    const dismiss = jest.fn()
    const wrapper = mount(<NotificationsPortal {...initialProps} store={modifiedStore} removeNotification={dismiss} />)
    wrapper.find(Notification).find('button').simulate('click');
    expect(dismiss).toHaveBeenCalledWith('store notification')
  });
});
