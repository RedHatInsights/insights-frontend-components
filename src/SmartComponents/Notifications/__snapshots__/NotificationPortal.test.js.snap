// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification portal should render notifications given as direct props 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    className="notifications-portal"
  >
    <Notification
      description="Some meaningfull description"
      dismissDelay={5000}
      dismissable={true}
      id="0"
      key="0"
      onDismiss={[Function]}
      title="Notification title"
      variant="success"
    />
  </div>
</Portal>
`;

exports[`Notification portal should render notifications given as direct props over store notifications 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    className="notifications-portal"
  >
    <Notification
      description="Some meaningfull description"
      dismissDelay={5000}
      dismissable={true}
      id="Direct notification"
      key="Direct notification"
      onDismiss={[Function]}
      title="Notification title"
      variant="success"
    />
  </div>
</Portal>
`;

exports[`Notification portal should render notifications given from store 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    className="notifications-portal"
  >
    <Notification
      description="Some meaningfull description"
      dismissDelay={5000}
      dismissable={true}
      id="0"
      key="0"
      onDismiss={[Function]}
      title="Notification title"
      variant="success"
    />
  </div>
</Portal>
`;
